File: lib/features/tickets/data/services/ticket_service.dart
import 'package:cal_ai/services/api_service.dart';
import 'package:cal_ai/features/tickets/data/models/ticket_model.dart';

class TicketService {
  final ApiService _apiService;

  TicketService(this._apiService);

  /// Create a new ticket
  Future<TicketModel> createTicket({
    required String subject,
    required String message,
    TicketCategory category = TicketCategory.general,
    TicketPriority priority = TicketPriority.medium,
  }) async {
    try {
      final response = await _apiService.post('/api/tickets', data: {
        'subject': subject,
        'message': message,
        'category': category.toServerString(),
        'priority': priority.toServerString(),
      });

      if (response['success'] == true && response['data'] != null) {
        return TicketModel.fromJson(
            response['data']['ticket'] as Map<String, dynamic>);
      }

      throw Exception(response['message'] ?? 'Failed to create ticket');
    } catch (e) {
      print('Error creating ticket: $e');
      rethrow;
    }
  }

  /// Get user's tickets
  Future<List<TicketModel>> getMyTickets({
    int page = 1,
    int limit = 20,
    TicketStatus? status,
  }) async {
    try {
      final queryParams = <String, dynamic>{
        'page': page.toString(),
        'limit': limit.toString(),
      };

      if (status != null) {
        queryParams['status'] = status.toServerString();
      }

      final response = await _apiService.get(
        '/api/tickets/my-tickets',
        queryParameters: queryParams,
      );

      if (response['success'] == true && response['data'] != null) {
        final ticketsData = response['data']['items'] as List<dynamic>;
        return ticketsData
            .map((json) => TicketModel.fromJson(json as Map<String, dynamic>))
            .toList();
      }

      return [];
    } catch (e) {
      print('Error fetching tickets: $e');
      return [];
    }
  }

  /// Get ticket by ID
  Future<TicketModel?> getTicketById(String ticketId) async {
    try {
      final response = await _apiService.get('/api/tickets/$ticketId');

      if (response['success'] == true && response['data'] != null) {
        return TicketModel.fromJson(
            response['data']['ticket'] as Map<String, dynamic>);
      }

      return null;
    } catch (e) {
      print('Error fetching ticket: $e');
      return null;
    }
  }

  /// Add a message to a ticket
  Future<TicketModel> addMessage({
    required String ticketId,
    required String message,
  }) async {
    try {
      final response = await _apiService.post(
        '/api/tickets/$ticketId/messages',
        data: {'message': message},
      );

      if (response['success'] == true && response['data'] != null) {
        return TicketModel.fromJson(
            response['data']['ticket'] as Map<String, dynamic>);
      }

      throw Exception(response['message'] ?? 'Failed to send message');
    } catch (e) {
      print('Error adding message: $e');
      rethrow;
    }
  }
}

File: panel/src/services/ticket.service.ts
import { API_BASE_URL, API_TIMEOUT } from '../config/api';
import { authService } from './auth.service';
import type { 
  Ticket, 
  PaginatedTickets, 
  TicketStatistics, 
  TicketStatus, 
  TicketPriority, 
  TicketCategory 
} from '../types/ticket';

class TicketService {
  private async fetchWithTimeout(url: string, options: RequestInit = {}) {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), API_TIMEOUT);

    try {
      const token = authService.getToken();
      const response = await fetch(url, {
        ...options,
        signal: controller.signal,
        headers: {
          'Content-Type': 'application/json',
          Authorization: token ? `Bearer ${token}` : '',
          ...options.headers,
        },
      });
      clearTimeout(timeout);
      return response;
    } catch (error) {
      clearTimeout(timeout);
      throw error;
    }
  }

  // Admin: List all tickets with filters
  async list(params: {
    page?: number;
    limit?: number;
    status?: TicketStatus;
    priority?: TicketPriority;
    category?: TicketCategory;
    userId?: string;
    search?: string;
    sort?: string;
  }): Promise<PaginatedTickets> {
    const query = new URLSearchParams();
    if (params.page) query.set('page', String(params.page));
    if (params.limit) query.set('limit', String(params.limit));
    if (params.status) query.set('status', params.status);
    if (params.priority) query.set('priority', params.priority);
    if (params.category) query.set('category', params.category);
    if (params.userId) query.set('userId', params.userId);
    if (params.search) query.set('search', params.search);
    if (params.sort) query.set('sort', params.sort);

    const url = `${API_BASE_URL}/api/tickets/admin/list?${query.toString()}`;
    const res = await this.fetchWithTimeout(url);
    return res.json();
  }

  // Get ticket by ID
  async getById(id: string): Promise<{ success: boolean; data?: { ticket: Ticket } }> {
    const url = `${API_BASE_URL}/api/tickets/${id}`;
    const res = await this.fetchWithTimeout(url);
    return res.json();
  }

  // Add message to ticket
  async addMessage(
    id: string, 
    message: string
  ): Promise<{ success: boolean; message?: string; data?: { ticket: Ticket } }> {
    const url = `${API_BASE_URL}/api/tickets/${id}/messages`;
    const res = await this.fetchWithTimeout(url, {
      method: 'POST',
      body: JSON.stringify({ message }),
    });
    return res.json();
  }

  // Update ticket status (admin only)
  async updateStatus(
    id: string, 
    status: TicketStatus,
    assignedTo?: string
  ): Promise<{ success: boolean; message?: string; data?: { ticket: Ticket } }> {
    const url = `${API_BASE_URL}/api/tickets/${id}/status`;
    const res = await this.fetchWithTimeout(url, {
      method: 'PUT',
      body: JSON.stringify({ status, assignedTo }),
    });
    return res.json();
  }

  // Update ticket priority (admin only)
  async updatePriority(
    id: string, 
    priority: TicketPriority
  ): Promise<{ success: boolean; message?: string; data?: { ticket: Ticket } }> {
    const url = `${API_BASE_URL}/api/tickets/${id}/priority`;
    const res = await this.fetchWithTimeout(url, {
      method: 'PUT',
      body: JSON.stringify({ priority }),
    });
    return res.json();
  }

  // Get ticket statistics (admin only)
  async getStatistics(): Promise<{ success: boolean; data?: TicketStatistics }> {
    const url = `${API_BASE_URL}/api/tickets/admin/statistics`;
    const res = await this.fetchWithTimeout(url);
    return res.json();
  }

  // Get unread ticket count (admin only)
  async getUnreadCount(): Promise<{ success: boolean; data?: { count: number } }> {
    const url = `${API_BASE_URL}/api/tickets/admin/unread-count`;
    const res = await this.fetchWithTimeout(url);
    return res.json();
  }

  // Delete ticket (admin only)
  async delete(id: string): Promise<{ success: boolean; message?: string }> {
    const url = `${API_BASE_URL}/api/tickets/${id}`;
    const res = await this.fetchWithTimeout(url, {
      method: 'DELETE',
    });
    return res.json();
  }
}

export const ticketService = new TicketService();
